### ISSUE TITLE
Is there any way to restart broken Enclave?

### ISSUE THREAD

**Issue by LeoneChen** [intel/linux-sgx#846]
When an exception occurs in Enclave, exception handler will be called. However, we want to close the broken enclave, and start a new enclave, it seems has no valid method. 1) If the exception make process killed, then we can do nothing. 2) If let exception handler return, then control flow go back to error position.

I know we can adjust rip address in in-Enclave handler registered in `sgx_register_exception_handler`, but it only available on SGXv2 when MiscSelect is open. I want a method also avaliable for SGXv1.

**Comment by LeoneChen** [comment]
Oh, I've found `CEnclave::rdunlock()`.

**Comment by lzha101** [comment]
> I know we can adjust rip address in in-Enclave handler registered in sgx_register_exception_handler, but it only available on SGXv2 when MiscSelect is open. I want a method also avaliable for SGXv1.

I don't think it needs SGX2 with MiscSelect enabled. You can register the handler for exception handling to avoid enclave crash for both SGX1 and SGX2. 

If your enclave is indeed broken for some reason, you can destroy it, recreate the enclave an resume the status by some backup status. Of cause, it requires you to backup your enclave status regularly. It depends on your enclave logic.

**Comment by LeoneChen** [comment]
Thanks for your reply~

> I don't think it needs SGX2 with MiscSelect enabled. You can register the handler for exception handling to avoid enclave crash for both SGX1 and SGX2.

Here is function `trts_handle_exception` in `sdk/trts/trts_veh.cpp` that will check `ssa_gpr->exit_info.valid`. 
```cpp
extern "C" sgx_status_t trts_handle_exception(void *tcs)
{
    ...
    if(ssa_gpr->exit_info.valid != 1)
    {   // exception handlers are not allowed to call in a non-exception state
        goto default_handler;
    }
    ...
    /* Some code to prepare the ip for 2nd phrase handling (i.e. internal_handle_exception) */
default_handler:
    g_enclave_state = ENCLAVE_CRASHED;
    return SGX_ERROR_ENCLAVE_CRASHED;
}
```
When I set `MiscSelect` to 0, `ssa_gpr->exit_info.valid` is 0, then goto  `default_handler`.

And `MiscSelect` can't set to 1 on SGXv1, otherwise will get error `SGX_ERROR_INVALID_MISC`. However, `MiscSelect` can be set to 1 on SGXv2 computer, and my registered in-Enclave handler(registered via `sgx_register_exception_handler` API) can be called as I expected.

Although in doc of SGXv2.16, it still say
![image](https://user-images.githubusercontent.com/35480440/172112729-d79b3121-da05-42f5-a33d-4c9888b1c48c.png)

**Comment by LeoneChen** [comment]
>If your enclave is indeed broken for some reason, you can destroy it, recreate the enclave an resume the status by some backup status. Of cause, it requires you to backup your enclave status regularly. It depends on your enclave logic.

When Enclave broken, I currently catch this at signal handler. If I restart enclave at this point, it's OK, but there are 3 options after that
1. abort process(the default operation), it's not what I want
2.  return, then it will go back to in-Enclave original error point, and since I've destroy enclave, then it will go to new broken status
3.  if we long jump from handler to e.g. main function in host app, it seems too hack

**Comment by lzha101** [comment]
May I know what code triggers the exception?  #PF?

According to [ Intel SDM](https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-vol-3d-part-4-manual.pdf),  

> VALID:
 0: unsupported exceptions.
1: Supported exceptions. Includes two categories:
• Unconditionally supported exceptions: #DE, #DB, #BP, #BR, #UD, #MF, #AC, #XM.
• Conditionally supported exception:
— #PF, #GP if SECS.MISCSELECT.EXINFO = 1.

So seems you happen to encounter the conditionally supported exception and makes your exception invalid on some platform.

**Comment by LeoneChen** [comment]
>May I know what code triggers the exception? #PF?

I'm currently fuzz sgx application, e.g. null-pointer dereference will trigger #PF, and SIGSEGV handler will catch this

>So seems you happen to encounter the conditionally supported exception and makes your exception invalid on some platform.

My SGXv1 PC's CPU is _Intel(R) Core(TM) i7-10700K CPU @ 3.80GHz_. 
```shell
$ cpuid -1 -l 0x12 -s 0x0 -r
CPU:
   0x00000012 0x00: eax=0x00000001 ebx=0x00000000 ecx=0x00000000 edx=0x0000241f
```

My SGXv2 server's CPU is _Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz_ 
```shell
$ cpuid -1 -l 0x12 -s 0x0 -r
CPU:
   0x00000012 0x00: eax=0x00000403 ebx=0x00000001 ecx=0x00000000 edx=0x0000381f
```

So, do you mean `#DE, #DB, #BP, #BR, #UD, #MF, #AC, #XM` will definitly set `ssa_gpr->exit_info.valid` to 1, while `#PF, #GP` will set `ssa_gpr->exit_info.valid` to 1 if and only if `SECS.MISCSELECT.EXINFO = 1`(which means `MiscSelect=1` in `Enclave.config.xml` and `CPUID.(EAX=12H, ECX=0):EBX[31:0] != 0`)

**Comment by lzha101** [comment]
Yes, except the CPUID.(EAX=12H, ECX=0):EBX[31:0] != 0. I think it only needs the EBX.bit0 is set to 1.

**Comment by LeoneChen** [comment]
OK, thanks for your reply~

**Comment by LeoneChen** [comment]
Hello~ Whether `sgx_destroy_enclave` is signal safe? If not, we can't call `sgx_destroy_enclave` in signal handler

**Comment by lzha101** [comment]
Do you want to destroy and restart an enclave in a signal handler outside if any exception happens inside enclave but not handled?  I don't think it is a good choice.  IMM,  the design of throwing a signal outside to crash the app is to simulate the behavior that an app dynamically loads a normal shared object and an unhandled exception is thrown from the shared object during execution.

And please note that after SGX 2.16 release, uRTS doesn't always throw and exception when the enclave throws an exception and doesn't register a handler to handle it inside enclave. It depends on whether your platform supports vDSO or not.  If your platform supports vDSO,  uRTS will directly return the error code of enclave crashed rather than throw the signal to app.  In this case, you just need to check the ecall return code to decide whether you want to destroy the enclave an recreate it.

**Comment by lzha101** [comment]
That's should not be true. OS will recycle all the resources including EPC when the app exits.

**Comment by LeoneChen** [comment]
@lzha101 Thanks! Any code to prove this? From driver?

**Comment by LeoneChen** [comment]
Dose process killed by SIGKILL not free EPC?

**Comment by LeoneChen** [comment]
Sorry, I found `sgx_release` in kernel will free all EPC. But it's weird to me why I fuzz (test target Enclave many many times) Enclave which may trigger signal handler and `_exit` without `sgx_destroy_enclave` cause DELL server crash and warning  `CPU0704: CPU1 machine check error detected`, I think opened file descriptors should be closed even at `_exit`, and then `sgx_release` will be called in kernel
